src/sentry/tasks/summaries/weekly_reports.py

```diff
@@ -331,41 +333,66 @@ def send_email(self, template_ctx: Mapping[str, Any], user_id: int) -> None:
-            extra = {
-                "batch_id": self.batch_id,
-                "organization": self.ctx.organization.id,
-                "user_id": user_id,
-                "has_email_override": bool(self.email_override),
-            }
-            logger.error(msg, extra=extra)
-
-        cluster = redis.clusters.get("default").get_local_client_for_key("weekly_reports")
-        name_parts = (self.batch_id, self.ctx.organization.id, user_id)
-        name = ":".join(str(part) for part in name_parts)
-
-        count_before = int(cluster.get(name) or 0)
-        if count_before > 0:
-            # When we have more confidence in this approach, we can upgrade this to
-            # an exception, thereby preventing the duplicate send.
-            log_error("weekly_report.delivery_record.duplicate_detected")
-
-        # Dispatch the send operation. There is no lock for concurrency, which leaves
-        # open the possibility of a race condition, in case another thread or server
-        # node received a duplicate Celery task somehow. But we do not think this is
-        # a likely failure mode.
-        yield
-
-        count_after = cluster.incr(name)
-        cluster.expire(name, timedelta(days=1))
-        if count_after > count_before + 1:
-            # The `cluster.incr` operation is atomic, so if concurrent duplicates are
-            # happening, this should reliably detect them after the fact.
-            log_error("weekly_report.delivery_record.concurrent_detected")
+class _DuplicateDeliveryCheck:
+    def __init__(self, batch: OrganizationReportBatch, user_id: int):
+        self.batch = batch
+        self.user_id = user_id
+
+        # Tracks state from `check_for_duplicate_delivery` to `record_delivery`
+        self.count: int | None = None
+
+    def _get_redis_cluster(self) -> LocalClient:
+        return redis.clusters.get("default").get_local_client_for_key("weekly_reports")
+
+    @property
+    def _redis_name(self) -> str:
+        name_parts = (self.batch.batch_id, self.batch.ctx.organization.id, self.user_id)
+        return ":".join(str(part) for part in name_parts)
+
+    def _get_log_extras(self) -> dict[str, Any]:
+        return {
+            "batch_id": self.batch.batch_id,
+            "organization": self.batch.ctx.organization.id,
+            "user_id": self.user_id,
+            "has_email_override": bool(self.batch.email_override),
+        }
+
+    def check_for_duplicate_delivery(self) -> bool:
+        """Check whether this delivery has been recorded in Redis already."""
+        if self.count is not None:
+            raise ValueError("This object has already checked a delivery")
+        cluster = self._get_redis_cluster()
+        self.count = int(cluster.get(self._redis_name) or 0)
+
+        is_duplicate_detected = self.count > 0
+        if is_duplicate_detected:
+            logger.error(
+                "weekly_report.delivery_record.duplicate_detected", extra=self._get_log_extras()
```

Comment:
Could we ever have more than a count of 1 here? It might be interesting to log the count but if it's always 1 here that's not very interesting data.

Reply:
Good idea -- we definitely could have a count of more than 1. I'll add it.